src/ui/mainwindow.cpp

```diff
@@ -358,6 +362,236 @@ void MainWindow::createStatusBar()
     scrollArea->setFixedHeight(frame->height());
 }
 
+
+bool MainWindow::saveSession(QString filePath, bool cacheModifiedFiles)
+{
+    std::vector<ViewData> viewData;
+
+    //Loop through all tabwidgets and their tabs
+    const int tabWidgetsCount = m_topEditorContainer->count();
+    for (int i = 0; i < tabWidgetsCount; i++) {
+        EditorTabWidget *tabWidget = m_topEditorContainer->tabWidget(i);
+        const int tabCount = tabWidget->count();
+
+        viewData.push_back( ViewData() );
+        ViewData& currentViewData = viewData.back();
+
+        for (int j = 0; j < tabCount; j++) {
+            Editor* editor = tabWidget->editor(j);
+            bool isClean = editor->isClean();
+            bool isOrphan = editor->fileName().isEmpty();
+
+            if (isOrphan && !cacheModifiedFiles)
+                continue; // Don't save temporary files if we're not caching tabs
+
+            TabData td;
+
+            if (!isClean && cacheModifiedFiles) {
+                // Tab is dirty, meaning it needs to be cached.
+                QUrl cacheFilePath = PersistentCache::createValidCacheName(tabWidget->tabText(j));
+
+                td.cacheFilePath = cacheFilePath.toLocalFile();
+
+                if (m_docEngine->saveDocument(tabWidget, j, cacheFilePath, true) != saveFileResult_Saved) {
+                    return false;
+                }
+            } else if (isOrphan) {
+                // Since we didn't cache the file and it is an orphan, we won't save it in the session.
+                continue;
+            }
+            // Else tab is an openened unmodified file, we don't have to do anything special.
+
+            td.filePath = !isOrphan ? editor->fileName().toLocalFile() : "" ;
+
+            // Finally save other misc information about the tab.
+            const auto& scrollPos = editor->scrollPosition();
+            td.scrollX = scrollPos.first;
+            td.scrollY = scrollPos.second;
+
+            // If we're caching and there's a file opened in the tab we want to inform the
+            // user whether the file's contents have changed since Nqq was last opened.
+            // For this we save and later compare the modification date.
+            if (!isOrphan && cacheModifiedFiles) {
+                // As a special case, if the file has *already* changed we set the modification
+                // time to 1 so we always trigger the warning.
+                if (editor->fileOnDiskChanged())
+                    td.lastModified = 1;
+                else
+                    td.lastModified = QFileInfo(td.filePath).lastModified().toMSecsSinceEpoch();
+            }
+
+            currentViewData.tabs.push_back( td );
+
+        } // end for
+    } // end for
+
+    // Write all information to a session file
+    QFile file(filePath);
+    file.open(QIODevice::WriteOnly);
+
+    if (file.isOpen()) {
+        SessionWriter sessionWriter(file);
```

Comment:
If we can't open the file in write mode we should report it to the caller!

Suppose, as an example, that I have no write permissions over my `~/.config/Notepadqq/session.xml`. I make some important changes without saving them because I think that notepadqq will save them in the session, but instead it will close without saying anything, losing my changes.

In this case, nqq should show a dialog saying that it's not able to save the session, and then abort the closing procedure thus giving the user a chance to manually save his files.